server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1509,85 @@ public void testSnapshotFromStreamInput() throws IOException {
         assertEquals(ops, readOperations);
     }
 
+    public void testTrimmedSnapshot() throws Exception {
+        List<Translog.Index> operations = new ArrayList<>();
+        int translogOperations = randomIntBetween(10, 100);
+        int maxTrimmedSeqNo = translogOperations - randomIntBetween(4, 8);
+
+        for (int op = 0; op < translogOperations; op++) {
+            String ascii = randomAlphaOfLengthBetween(1, 50);
+            Translog.Index operation = new Translog.Index("test", "" + op, op,
+                primaryTerm.get(), ascii.getBytes("UTF-8"));
+            operations.add(operation);
+        }
+        // shuffle a bit - move several first items to the end
+        for(int i = 0, len = randomIntBetween(5, 10); i < len; i++){
+            operations.add(operations.remove(0));
+        }
+
+        for (Translog.Index operation : operations) {
+            translog.add(operation);
+        }
+
+        translog.rollGeneration();
+
+        translog.trim(primaryTerm.get() + 1, maxTrimmedSeqNo);
+        translog.sync();
+
+        Set<Long> expectedSeqNo = LongStream.range(0, maxTrimmedSeqNo + 1).boxed().collect(Collectors.toSet());
+        try (Translog.Snapshot snapshot = translog.newSnapshot()) {
+            Translog.Operation next;
+            while ((next = snapshot.next()) != null) {
+                assertThat(expectedSeqNo.remove(next.seqNo()), is(true));
+            }
+            assertThat(snapshot.skippedOperations(), is(translogOperations - (maxTrimmedSeqNo + 1)));
+        }
+        assertThat(expectedSeqNo, hasSize(0));
+    }
+
+    public void testExceptionOnTrimAboveSeqNo( ) throws Exception {
+        Path tempDir = createTempDir();
+        final FailSwitch fail = new FailSwitch();
+        fail.failNever();
+        TranslogConfig config = getTranslogConfig(tempDir);
+        final Translog failableTLog = getFailableTranslog(fail, config, randomBoolean(), false, null, createTranslogDeletionPolicy());
+
+        List<Translog.Index> operations = new ArrayList<>();
+        int translogOperations = randomIntBetween(10, 100);
+        int maxTrimmedSeqNo = translogOperations - randomIntBetween(4, 8);
+
+        for (int op = 0; op < translogOperations; op++) {
+            String ascii = randomAlphaOfLengthBetween(1, 50);
+            Translog.Index operation = new Translog.Index("test", "" + op, op,
+                primaryTerm.get(), ascii.getBytes("UTF-8"));
+            operations.add(operation);
+        }
+        // shuffle a bit - move several first items to the end
+        for(int i = 0, len = randomIntBetween(5, 10); i < len; i++){
+            operations.add(operations.remove(0));
+        }
+
+        for (Translog.Index operation : operations) {
+            failableTLog.add(operation);
+        }
+
+        failableTLog.rollGeneration();
```

Comment:
as in the previous test - you need way more evilness - more roll generations, more terms, move variance.